packages/dds/tree/src/brand.ts

```diff
@@ -0,0 +1,108 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Invariant } from "./typeCheck";
+
+/**
+ * Constructs a "Branded" type, adding a type-checking only field to `ValueType`.
+ *
+ * Two usages of `Brand` should never use the same `Name`.
+ * If they do, the resulting types will be assignable which defeats the point of this type.
+ *
+ * This type is constructed such that the first line of type errors when assigning mismatched branded types will be:
+ * `Type 'Name1' is not assignable to type 'Name2'.`
+ *
+ * These branded types are not opaque: A `Brand<A, B>` can still be used as a `B`.
+ */
+export type Brand<ValueType, Name extends string> = ValueType &
+    BrandedType<ValueType, Name>;
+
+/**
+ * Helper for {@link Brand}.
+ * This is split out into its own as thats the only way to:
+ * - have doc comments for the field.
+ * - make the field protected (so you don't accidentally try and read it).
+ * - get nominal typing (so types produces without using this class can never ne assignable to it).
```

Comment:
two typos in one line, I'm obliged to comment (my current PR typo tolerance configuration is set to only allow one typo per line)